gapis/gfxapi/gles/gles.api

```diff
@@ -515,34 +515,47 @@ sub void ApplyStaticContextState(ref!Context ctx, ref!StaticContextState staticS
 
 sub void ApplyDynamicContextState(ref!Context ctx, ref!DynamicContextState dynamicState) {
   if (dynamicState != null) {
-    backbuffer := ctx.Objects.Framebuffers[0]
-    color_id := backbuffer.ColorAttachments[0].Renderbuffer.ID
-    depth_id := backbuffer.DepthAttachment.Renderbuffer.ID
-    stencil_id := backbuffer.StencilAttachment.Renderbuffer.ID
-
-    color_buffer := ctx.SharedObjects.Renderbuffers[color_id]
-    color_buffer.Width = dynamicState.BackbufferWidth
-    color_buffer.Height = dynamicState.BackbufferHeight
-    color_buffer.InternalFormat = dynamicState.BackbufferColorFmt
-    ctx.SharedObjects.Renderbuffers[color_id] = color_buffer
-
-    depth_buffer := ctx.SharedObjects.Renderbuffers[depth_id]
-    depth_buffer.Width = dynamicState.BackbufferWidth
-    depth_buffer.Height = dynamicState.BackbufferHeight
-    depth_buffer.InternalFormat = dynamicState.BackbufferDepthFmt
-    ctx.SharedObjects.Renderbuffers[depth_id] = depth_buffer
-
-    stencil_buffer := ctx.SharedObjects.Renderbuffers[stencil_id]
-    stencil_buffer.Width = dynamicState.BackbufferWidth
-    stencil_buffer.Height = dynamicState.BackbufferHeight
-    stencil_buffer.InternalFormat = dynamicState.BackbufferStencilFmt
-    ctx.SharedObjects.Renderbuffers[stencil_id] = stencil_buffer
-
-    if dynamicState.ResetViewportScissor {
-      ctx.FragmentOperations.Scissor.Box.Width = dynamicState.BackbufferWidth
-      ctx.FragmentOperations.Scissor.Box.Height = dynamicState.BackbufferHeight
-      ctx.Rasterization.Viewport.Width = dynamicState.BackbufferWidth
-      ctx.Rasterization.Viewport.Height = dynamicState.BackbufferHeight
+    if 0 in ctx.Objects.Framebuffers {
```

Comment:
Putting the offline discussion here so that it is public:

> With `eglMakeCurrent` missing this would just result in no context being bound, which would make other commands gracefully stop.

Okay, so the failure probably isn't with a preceding `eglMakeCurrent`, but more likely `glGenFramebuffers` reading zeros (due to cancellation). As we get back zeros, we're overwriting the 0'th framebuffer with a newly initialized one, corrupting the state. As discussed, I'll try adding an assert that the read value is not 0, and run my tests again. If this is the case, I'm happy to substitute the first CL with this fix.

> I say you should finish the first one, skip the 998 which have not even started yet, and then do the last one.

So we're now running all state mutations on a single executor? You know there's about 15 state mutations run concurrently whenever a command is clicked? Even if we wanted to go down this path (which I would suspect would make things much slower), adding a queued executor sounds like a few weeks work.

> If you really want to try hard, I say you can check for cancellation in the top-most loop so you bail out once then next atom is fully processed.

That's where the code does exit the loop. I have not changed things so that the `Mutate()` call will return any earlier than it used to. The template changes do two things:

1. Panics if we get an error back that isn't `nil` or `context.Canceled`. It completely ignores `context.Canceled` if it sees it.
2. Returns `ctx.Err()` instead of `nil` at the end of the function. This is just so the callee of `Mutate` doesn't have to check both the `error` returned by `Mutate` and `ctx.Err()` to know whether it should stop.

In terms of state inconsistency - there is a small window in which we can get unexpected data. If the context is cancelled between the `Mutate` start and end, you'll get undefined data on memory reads **for that command only**. After that the `Mutate` function will return `context.Canceled` and the callee should abort the loop. In this situation the state should not be used as the request was cancelled - stop whatever your doing and return.

> I fail to see why it is important to cancel immediately in the most nested function and then deal with the error propagation and state inconsistency problems.

As I've tried to explain, `context.Context` is designed so that cancellation propagates down to all child `Context`s. There is no way to stop a child context being cancelled when the parent is cancelled, unless you build an entirely new context, and copy over the content. This is infeasibly slow, and will lose any data added to the context by non-gapid packages (say gRPC).
`Mutate()` calls into other packages. Some of these packages respect cancellation if they are blocking or just take a long time. Things like database reads respect cancellation, and return the `context.Canceled` error when the callee context says it should stop.

I understand that handling cancellation mid-execution is a PITA, and not something we want to do, so instead of building "yet another really complicated solution", I tried doing the simplest solution - ignore the cancellation inside the `Mutate`, tell the callee if a cancellation occurred.

Please, if you have a better / less-complicated solution, share it.